pkg/ccl/streamingccl/streamingest/node_lag_detector_test.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package streamingest
+
+import (
+	"testing"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/server/status/statuspb"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/randutil"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/stretchr/testify/require"
+)
+
+func TestDetectLaggingNode(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	rng, _ := randutil.NewTestRand()
+
+	detail := func(srcID, dstID, frontier int) frontierExecutionDetails {
+		return frontierExecutionDetails{
+			srcInstanceID:  base.SQLInstanceID(srcID),
+			destInstanceID: base.SQLInstanceID(dstID),
+			frontierTS:     hlc.Timestamp{WallTime: int64(frontier)},
+		}
+	}
+
+	makeDetails := func(details ...frontierExecutionDetails) []frontierExecutionDetails {
+		for i := len(details) - 1; i > 0; i-- {
+			j := rng.Intn(i + 1)
+			details[i], details[j] = details[j], details[i]
+		}
+		return details
+	}
+
+	for _, tc := range []struct {
+		name             string
+		details          []frontierExecutionDetails
+		laggingNode      base.SQLInstanceID
+		minLagDifference time.Duration
+	}{
+		{
+			name: "simple",
+			details: makeDetails(
+				detail(1, 1, 3),
+				detail(2, 2, 1),
+				detail(3, 3, 6)),
+			laggingNode:      2,
+			minLagDifference: 2,
+		},
+		{
+			name: "aggregateOnLagger",
+			details: makeDetails(
+				detail(1, 1, 4),
+				detail(2, 2, 1),
+				detail(3, 2, 5),
+				detail(4, 3, 5)),
+			laggingNode:      2,
+			minLagDifference: 3,
+		},
+		{
+			name: "aggregateOnBoth",
+			details: makeDetails(
+				detail(1, 1, 1),
+				detail(2, 1, 6),
+				detail(3, 2, 2),
+				detail(4, 2, 4)),
+
+			laggingNode:      1,
+			minLagDifference: 1,
+		},
+		{
+			name: "skipInitialScanNode",
+			details: makeDetails(
+				detail(1, 1, 0),
+				detail(2, 2, 3),
+				detail(3, 2, 2),
+				detail(4, 3, 4)),
+
+			laggingNode:      2,
+			minLagDifference: 2,
+		},
+		{
```

Comment:
I'm unsure if the laggingNodeDetector should target a node finishing up an initial scan, or if the detector should even run if there's an initial scan taking place. Thoughts? Example: in the test case below, we select node 1 as the lagging node even though it is undergoing an initial scan.